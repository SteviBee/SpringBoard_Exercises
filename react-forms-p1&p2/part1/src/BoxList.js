import React, { useState } from "react";
import NewBoxForm from "./NewBoxForm";
import Box from "./Box";
import { v4 as uuid } from "uuid";

const BoxList = () => {
    // Set state initally to an empty array, state is an array of objs in the list
    const [items, setItems] = useState([])

    // Create Remove function:
    const removeBox = evt => {
        // Shift Opt F = to auto space
        let answer = items.filter(box => box.id !== evt.target.parentElement.id)
        setItems(answer)

    }

    // Create Rendering function: return new array with all the item parameters
    const renderItems = () => {
        return (        
            <div className="BoxList-Boxes">
                {items.map(item => (
                    <Box key={item.id} id={item.id} width={item.width} height={item.height} bgColor={item.bgColor} removeBox={removeBox}></Box>
               
                ))}
            </div>      
        )
    }


    /** Add new item object to boxlist. */
    const addItem = item => {
        let newItem = { ...item, id: uuid() };
        setItems(items => [...items, newItem]);
    };

    // Return everything
    return (
        <div className="BoxList">
            <NewBoxForm addItem={addItem}/>
            {renderItems()}

        </div>
    )

}


export default BoxList;