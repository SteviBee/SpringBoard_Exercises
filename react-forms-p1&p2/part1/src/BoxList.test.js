import React from "react";
import { render, fireEvent } from "@testing-library/react";
import BoxList from "./BoxList";
// NOTE - 5/18 - had to install jest-dom via npm install jest-dom
// Also imported it here
import '@testing-library/jest-dom'

it("renders withoutcrashing", function () {
    render(<BoxList />)
})

it('matches snapshot', () => {
    const { asFragment } = render(<BoxList />)
    expect(asFragment()).toMatchSnapshot()
});

// Critical business logic tested here: that the form displays, renders, and removes
it('should add a new box', () => {
    const { queryByText, getByLabelText } = render(<BoxList/>)
    const inputW = getByLabelText("Width:")
    const inputH = getByLabelText("Height:")
    const inputBC = getByLabelText("bgColor:")
    const btn = queryByText("Add Box");
    console.log(queryByText("X"))

    expect(queryByText('X')).not.toBeInTheDocument();

    fireEvent.change(inputW, { target: { value: '100' } });
    fireEvent.change(inputH, { target: { value: '200' } });
    fireEvent.change(inputBC, { target: { value: 'green' } });
    fireEvent.click(btn);
    expect(queryByText('X')).toBeInTheDocument();
});

it('should REMOVE a new box', () => {
    const { queryByText, getByLabelText } = render(<BoxList/>)
    const inputW = getByLabelText("Width:")
    const inputH = getByLabelText("Height:")
    const inputBC = getByLabelText("bgColor:")
    const btn = queryByText("Add Box");
    console.log(queryByText("X"))

    expect(queryByText('X')).not.toBeInTheDocument();

    fireEvent.change(inputW, { target: { value: '100' } });
    fireEvent.change(inputH, { target: { value: '200' } });
    fireEvent.change(inputBC, { target: { value: 'green' } });
    fireEvent.click(btn);
    expect(queryByText('X')).toBeInTheDocument();

    // Now remove that box
    const remove = queryByText("X")
    fireEvent.click(remove);
    expect(queryByText('X')).not.toBeInTheDocument();
});

