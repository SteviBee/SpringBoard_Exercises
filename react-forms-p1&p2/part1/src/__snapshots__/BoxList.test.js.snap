// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="BoxList"
  >
    <form>
      <label
        for="width"
      >
        Width:
      </label>
      <input
        id="width"
        name="width"
        type="text"
        value="100"
      />
      <label
        for="height"
      >
        Height: 
      </label>
      <input
        id="height"
        max="10"
        min="1"
        name="height"
        type="text"
        value="100"
      />
      <label
        for="bgColor"
      >
        bgColor: 
      </label>
      <input
        id="bgColor"
        max="10"
        min="1"
        name="bgColor"
        type="text"
        value="black"
      />
      <button>
        Add Box
      </button>
    </form>
    <div
      class="BoxList-Boxes"
    />
  </div>
</DocumentFragment>
`;
